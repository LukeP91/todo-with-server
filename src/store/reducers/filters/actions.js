import { types } from './types'

export const setFilter = value => {
  return {
    type: types.SET_FILTER,
    payload: value,
  }
}

export const setVisiblityFilter = value => {
  return {
    type: types.SET_VISIBILITY_FILTER,
    payload: value,
  }
}
