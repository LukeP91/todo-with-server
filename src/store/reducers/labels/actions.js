import { types } from './types'

const fetchLabelsRequest = () => {
  return {
    type: types.FETCH_LABELS_REQUEST,
  }
}

const fetchLabelsSuccess = payload => {
  return {
    type: types.FETCH_LABELS_SUCCESS,
    payload,
  }
}

const fetchLabelsError = () => {
  return {
    type: types.FETCH_LABELS_ERROR,
  }
}

const fetchData = (url, params = {}) => {
  return fetch(url, params).then(response => Promise.all([response, response.json()]))
}

export const fetchLabels = (url, params) => {
  return dispatch => {
    dispatch(fetchLabelsRequest())
    return fetchData(url, params).then(([response, json]) => {
      if (response.status === 200) {
        dispatch(fetchLabelsSuccess(json))
      } else {
        dispatch(fetchLabelsError())
      }
    })
  }
}
