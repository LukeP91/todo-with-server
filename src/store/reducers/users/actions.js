import { types } from './types'

const fetchUsersRequest = () => {
  return {
    type: types.FETCH_USERS_REQUEST,
  }
}

const fetchUsersSuccess = payload => {
  return {
    type: types.FETCH_USERS_SUCCESS,
    payload,
  }
}

const fetchUsersError = () => {
  return {
    type: types.FETCH_USERS_ERROR,
  }
}

const fetchData = (url, params = {}) => {
  return fetch(url, params).then(response => Promise.all([response, response.json()]))
}

export const fetchUsers = (url, params) => {
  return dispatch => {
    dispatch(fetchUsersRequest())
    return fetchData(url, params).then(([response, json]) => {
      if (response.status === 200) {
        dispatch(fetchUsersSuccess(json))
      } else {
        dispatch(fetchUsersError())
      }
    })
  }
}
