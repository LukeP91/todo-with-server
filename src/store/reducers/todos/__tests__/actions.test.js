import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import fetchMock from 'fetch-mock'

import * as actions from '../actions'

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

describe('actionCreators', () => {
  afterEach(() => {
    fetchMock.reset()
    fetchMock.restore()
  })

  it('creates FETCH_TODOS_SUCCESS when fetching todos has been successful', () => {
    fetchMock.getOnce('http://example.com/todos', {
      body: {
        todos: [{ description: 'new todo', completed: false, id: 1 }],
      },
    })

    const store = mockStore({ todos: [] })

    return store
      .dispatch(actions.fetchTodos('http://example.com/todos', { method: 'GET' }))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot()
      })
  })

  it('creates REMOVE_TODO_SUCCESS when removing todo has been successful', () => {
    fetchMock.deleteOnce('http://example.com/todos/1', {})

    const store = mockStore({ todos: [] })

    return store
      .dispatch(actions.removeTodo('http://example.com/todos/1', { method: 'DELETE' }))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot()
      })
  })

  it('creates ADD_TODO_SUCCESS when adding todo has been successful', () => {
    fetchMock.postOnce('http://example.com/todos', {
      body: { description: 'new todo', completed: false, id: 1, userId: 1, labelsIds: [1, 2] },
      status: 201,
    })

    const store = mockStore({ todos: [] })

    return store
      .dispatch(
        actions.addTodo('http://example.com/todos', {
          method: 'POST',
          headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            description: 'new todo',
            completed: false,
            userId: 1,
            labelsIds: [1, 2],
          }),
        }),
      )
      .then(() => {
        expect(store.getActions()).toMatchSnapshot()
      })
  })

  it('creates TOGGLE_TODO_SUCCESS when removing todo has been successful', () => {
    fetchMock.patchOnce('http://example.com/todos/1', {})

    const store = mockStore({ todos: [] })

    return store
      .dispatch(actions.toggleTodo('http://example.com/todos/1', { method: 'PATCH' }))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot()
      })
  })

  it('creates EDIT_TODO_SUCCESS when editing todo has been successful', () => {
    fetchMock.patchOnce('http://example.com/todos/1', {
      body: { id: 1, name: 'new name' },
    })

    const store = mockStore({ todos: [{ id: 1, name: 'old name', description: 'desc' }] })

    return store
      .dispatch(
        actions.editTodo('http://example.com/todos/1', {
          method: 'PATCH',
          body: { id: 1, name: 'new name' },
        }),
      )
      .then(() => {
        expect(store.getActions()).toMatchSnapshot()
      })
  })
})
