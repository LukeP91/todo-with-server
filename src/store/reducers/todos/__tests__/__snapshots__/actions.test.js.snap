// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actionCreators creates ADD_TODO_SUCCESS when adding todo has been successful 1`] = `
Array [
  Object {
    "type": "ADD_TODO_REQUEST",
  },
  Object {
    "payload": Object {
      "completed": false,
      "description": "new todo",
      "id": 1,
      "labelsIds": Array [
        1,
        2,
      ],
      "userId": 1,
    },
    "type": "ADD_TODO_SUCCESS",
  },
]
`;

exports[`actionCreators creates EDIT_TODO_SUCCESS when editing todo has been successful 1`] = `
Array [
  Object {
    "type": "EDIT_TODO_REQUEST",
  },
  Object {
    "payload": Object {
      "id": 1,
      "name": "new name",
    },
    "type": "EDIT_TODO_SUCCESS",
  },
]
`;

exports[`actionCreators creates FETCH_TODOS_SUCCESS when fetching todos has been successful 1`] = `
Array [
  Object {
    "type": "FETCH_TODOS_REQUEST",
  },
  Object {
    "payload": Object {
      "todos": Array [
        Object {
          "completed": false,
          "description": "new todo",
          "id": 1,
        },
      ],
    },
    "type": "FETCH_TODOS_SUCCESS",
  },
]
`;

exports[`actionCreators creates REMOVE_TODO_SUCCESS when removing todo has been successful 1`] = `
Array [
  Object {
    "type": "REMOVE_TODO_REQUEST",
  },
  Object {
    "payload": 1,
    "type": "REMOVE_TODO_SUCCESS",
  },
]
`;

exports[`actionCreators creates TOGGLE_TODO_SUCCESS when removing todo has been successful 1`] = `
Array [
  Object {
    "type": "TOGGLE_TODO_REQUEST",
  },
  Object {
    "payload": 1,
    "type": "TOGGLE_TODO_SUCCESS",
  },
]
`;
